Rust build fails for target "aarch64-wrs-vxworks"

<!--
Thank you for submitting a bootstrap report! Please provide detailed information to help us reproduce and diagnose the issue.
-->

### Summary

Attempting to perform bootstrap rust build for target "aarch64-wrs-vxworks".

<!--
Provide a brief description of the problem you are experiencing.
-->

### Command used

```sh
./x.py build target=x86_64-unknown-linux-gnu,aarch64-wrs-vxworks
```

### Expected behaviour

<!--
Describe what you expected to happen.
-->
The rust build passes without any failures.

### Actual behaviour

<!--
Describe what actually happened.
-->

### Bootstrap configuration (bootstrap.toml)
```toml
llvm.download-ci-llvm = true
build.optimized-compiler-builtins = false
```

### Operating system

<!--
e.g., Ubuntu 22.04, macOS 12, Windows 10
-->
VxWorks

### HEAD

<!--
Output of `git rev-parse HEAD` command, or content of the `git-commit-hash` file if using a tarball source.
-->
d3877ec33269edb7355ef8538d1cc5d157aa1ace

### Additional context
<!--
Include any other relevant information (e.g., if you have custom patches or modifications on the project).
-->
The failure is only seen for vxworks version greater than or equal to 25.09.

<!--
Include the complete build log in the section below.
Enable backtrace and verbose mode if possible for more detailed information e.g., with `RUST_BACKTRACE=1 ./x build -v`.
-->
<details><summary>Build Log</summary>
<p>

```txt
Building stage1 library artifacts{alloc, compiler_builtins, core, panic_abort, panic_unwind, proc_macro, rustc-std-workspace-core, std, std_detect, sysroot, test, unwind} (stage1:x86_64-unknown-linux-gnu -> stage1:aarch64-wrs-vxworks)
   Compiling compiler_builtins v0.1.160 (<path>/rust/library/compiler-builtins/compiler-builtins)
   Compiling core v0.0.0 (<path>/rust/library/core)
   Compiling libc v0.2.178
   Compiling object v0.37.3
   Compiling std v0.0.0 (<path>/rust/library/std)
   Compiling test v0.0.0 (<path>/rust/library/test)
   Compiling rustc-std-workspace-core v1.99.0 (<path>/rust/library/rustc-std-workspace-core)
   Compiling alloc v0.0.0 (<path>/rust/library/alloc)
   Compiling memchr v2.7.6
   Compiling adler2 v2.0.1
   Compiling cfg-if v1.0.4
   Compiling panic_abort v0.0.0 (<path>/rust/library/panic_abort)
   Compiling rustc-demangle v0.1.27
   Compiling rustc-literal-escaper v0.0.7
   Compiling unwind v0.0.0 (<path>/rust/library/unwind)
   Compiling rustc-std-workspace-alloc v1.99.0 (<path>/rust/library/rustc-std-workspace-alloc)
   Compiling panic_unwind v0.0.0 (<path>/rust/library/panic_unwind)
   Compiling gimli v0.32.3
   Compiling std_detect v0.1.5 (<path>/rust/library/std_detect)
   Compiling hashbrown v0.16.1
   Compiling miniz_oxide v0.8.9
   Compiling addr2line v0.25.1
error[E0609]: no field `st_atime` on type `&stat`
  --> library/std/src/os/vxworks/fs.rs:73:36
   |
73 |         self.as_inner().as_inner().st_atime as i64
   |                                    ^^^^^^^^ unknown field
   |
help: a field with a similar name exists
   |
73 -         self.as_inner().as_inner().st_atime as i64
73 +         self.as_inner().as_inner().st_atim as i64
   |

error[E0609]: no field `st_mtime` on type `&stat`
  --> library/std/src/os/vxworks/fs.rs:79:36
   |
79 |         self.as_inner().as_inner().st_mtime as i64
   |                                    ^^^^^^^^ unknown field
   |
help: a field with a similar name exists
   |
79 -         self.as_inner().as_inner().st_mtime as i64
79 +         self.as_inner().as_inner().st_mtim as i64
   |

error[E0609]: no field `st_ctime` on type `&stat`
  --> library/std/src/os/vxworks/fs.rs:85:36
   |
85 |         self.as_inner().as_inner().st_ctime as i64
   |                                    ^^^^^^^^ unknown field
   |
help: a field with a similar name exists
   |
85 -         self.as_inner().as_inner().st_ctime as i64
85 +         self.as_inner().as_inner().st_ctim as i64
   |

error[E0609]: no field `st_mtime` on type `stat`
   --> library/std/src/sys/fs/unix.rs:643:35
    |
643 |         SystemTime::new(self.stat.st_mtime as i64, 0)
    |                                   ^^^^^^^^ unknown field
    |
help: a field with a similar name exists
    |
643 -         SystemTime::new(self.stat.st_mtime as i64, 0)
643 +         SystemTime::new(self.stat.st_mtim as i64, 0)
    |

error[E0609]: no field `st_atime` on type `stat`
   --> library/std/src/sys/fs/unix.rs:678:35
    |
678 |         SystemTime::new(self.stat.st_atime as i64, 0)
    |                                   ^^^^^^^^ unknown field
    |
help: a field with a similar name exists
    |
678 -         SystemTime::new(self.stat.st_atime as i64, 0)
678 +         SystemTime::new(self.stat.st_atim as i64, 0)
    |

warning: unnecessary `unsafe` block
   --> library/std/src/sys/fd/unix.rs:165:13
    |
165 |         cvt(unsafe {
    |             ^^^^^^ unnecessary `unsafe` block
    |
    = note: `#[warn(unused_unsafe)]` (part of `#[warn(unused)]`) on by default

warning: unnecessary `unsafe` block
   --> library/std/src/sys/fd/unix.rs:407:9
    |
407 |         unsafe {
    |         ^^^^^^ unnecessary `unsafe` block

For more information about this error, try `rustc --explain E0609`.
warning: `std` (lib) generated 2 warnings
error: could not compile `std` (lib) due to 5 previous errors; 2 warnings emitted
```

</p>
</details>

--- a/library/std/build.rs
+++ b/library/std/build.rs
@@ -79,4 +79,28 @@ fn main() {
     println!("cargo:rustc-cfg=backtrace_in_libstd");
 
     println!("cargo:rustc-env=STD_ENV_ARCH={}", env::var("CARGO_CFG_TARGET_ARCH").unwrap());
+
+    println!("cargo:rustc-check-cfg=cfg(vxworks_lt_25_09)");
+
+    if target_os == "vxworks" {
+        match vxworks_version_code() {
+            Some((major, minor)) if (major, minor) < (25, 9) => {
+                println!("cargo:rustc-cfg=vxworks_lt_25_09");
+            }
+            _ => {}
+        }
+    }
+}
+
+/// Retrieve the VxWorks release version from the environment variable set by the VxWorks build
+/// environment, in `(minor, patch)` form.
+fn vxworks_version_code() -> Option<(u32, u32)> {
+    let version = env::var("WIND_RELEASE_ID").ok()?;
+
+    let mut pieces = version.trim().split(['.']);
+
+    let major: u32 = pieces.next().and_then(|x| x.parse().ok()).unwrap_or(0);
+    let minor: u32 = pieces.next().and_then(|x| x.parse().ok()).unwrap_or(0);
+
+    Some((major, minor))
 }

--- a/library/std/src/os/vxworks/fs.rs
+++ b/library/std/src/os/vxworks/fs.rs
@@ -69,23 +69,53 @@ impl MetadataExt for Metadata {
     fn st_size(&self) -> u64 {
         self.as_inner().as_inner().st_size as u64
     }
+    #[cfg(vxworks_lt_25_09)]
     fn st_atime(&self) -> i64 {
         self.as_inner().as_inner().st_atime as i64
     }
+    #[cfg(not(vxworks_lt_25_09))]
+    fn st_atime(&self) -> i64 {
+        self.as_inner().as_inner().st_atim.tv_sec as i64
+    }
+    #[cfg(vxworks_lt_25_09)]
     fn st_atime_nsec(&self) -> i64 {
         0
     }
+    #[cfg(not(vxworks_lt_25_09))]
+    fn st_atime_nsec(&self) -> i64 {
+        self.as_inner().as_inner().st_atim.tv_nsec as i64
+    }
+    #[cfg(vxworks_lt_25_09)]
     fn st_mtime(&self) -> i64 {
         self.as_inner().as_inner().st_mtime as i64
     }
+    #[cfg(not(vxworks_lt_25_09))]
+    fn st_mtime(&self) -> i64 {
+        self.as_inner().as_inner().st_mtim.tv_sec as i64
+    }
+    #[cfg(vxworks_lt_25_09)]
     fn st_mtime_nsec(&self) -> i64 {
         0
     }
+    #[cfg(not(vxworks_lt_25_09))]
+    fn st_mtime_nsec(&self) -> i64 {
+        self.as_inner().as_inner().st_mtim.tv_nsec as i64
+    }
+    #[cfg(vxworks_lt_25_09)]
     fn st_ctime(&self) -> i64 {
         self.as_inner().as_inner().st_ctime as i64
     }
+    #[cfg(not(vxworks_lt_25_09))]
+    fn st_ctime(&self) -> i64 {
+        self.as_inner().as_inner().st_ctim.tv_sec as i64
+    }
+    #[cfg(vxworks_lt_25_09)]
     fn st_ctime_nsec(&self) -> i64 {
         0
+    }
+    #[cfg(not(vxworks_lt_25_09))]
+    fn st_ctime_nsec(&self) -> i64 {
+        self.as_inner().as_inner().st_ctim.tv_nsec as i64
     }
     fn st_blksize(&self) -> u64 {
         self.as_inner().as_inner().st_blksize as u64

--- a/library/std/src/sys/fs/unix.rs
+++ b/library/std/src/sys/fs/unix.rs
@@ -634,7 +634,7 @@ impl FileAttr {
     }
 
     #[cfg(any(
-        target_os = "vxworks",
+        all(target_os = "vxworks", vxworks_lt_25_09),
         target_os = "espidf",
         target_os = "vita",
         target_os = "rtems",
@@ -643,7 +643,12 @@ impl FileAttr {
         SystemTime::new(self.stat.st_mtime as i64, 0)
     }
 
-    #[cfg(any(target_os = "horizon", target_os = "hurd", target_os = "nuttx"))]
+    #[cfg(any(
+        target_os = "horizon",
+        target_os = "hurd",
+        target_os = "nuttx",
+        all(target_os = "vxworks", not(vxworks_lt_25_09))
+    ))]
     pub fn modified(&self) -> io::Result<SystemTime> {
         SystemTime::new(self.stat.st_mtim.tv_sec as i64, self.stat.st_mtim.tv_nsec as i64)
     }
@@ -669,7 +674,7 @@ impl FileAttr {
     }
 
     #[cfg(any(
-        target_os = "vxworks",
+        all(target_os = "vxworks", vxworks_lt_25_09),
         target_os = "espidf",
         target_os = "vita",
         target_os = "rtems"
@@ -678,7 +683,12 @@ impl FileAttr {
         SystemTime::new(self.stat.st_atime as i64, 0)
     }
 
-    #[cfg(any(target_os = "horizon", target_os = "hurd", target_os = "nuttx"))]
+    #[cfg(any(
+        target_os = "horizon",
+        target_os = "hurd",
+        target_os = "nuttx",
+        all(target_os = "vxworks", not(vxworks_lt_25_09))
+    ))]
     pub fn accessed(&self) -> io::Result<SystemTime> {
         SystemTime::new(self.stat.st_atim.tv_sec as i64, self.stat.st_atim.tv_nsec as i64)
     }

--- a/src/doc/rustc/src/platform-support/vxworks.md
+++ b/src/doc/rustc/src/platform-support/vxworks.md
@@ -28,6 +28,12 @@ Target triplets available:
 
 The minimum supported version is VxWorks 7.
 
+### Environment
+
+#### `WIND_RELEASE_ID`
+
+In VxWorks build environment, the environment variable `WIND_RELEASE_ID` indicates the VxWorks release version used for the build. The `WIND_RELEASE_ID` can be used to conditionally compile features/code or handle version specific behaviour.
+
 ## Building
 
 Rust for each target can be cross-compiled with its specific target vsb configuration. Std support is added but not yet fully tested.

